### DISCUSSION TITLE
build questions

### DISCUSSION THREAD

**Discussion by ftc2** [userdocs/qbittorrent-nox-static#116] [General]
first of all – pretty cool script. thanks a lot.

i have some questions. i'm running this right now:

```bash
docker run -it -w /root -v $HOME/qbt:/root alpine:edge /bin/ash -c 'apk update && apk add bash curl && curl -sL git.io/qbstatic | \ 
bash -s all --qbittorrent-master --optimize --icu'
```

basically i want a musl bin for qbittorrent for an unreleased feature that's out in master that i'm very eager to try. hopefully i don't need master of libtorrent too for best results.

after reading `--help` and https://userdocs.github.io/qbittorrent-nox-static/#/script-installation?id=docker-via-ssh

`--optimize` is clear. i get that it's not good if i distribute it or use it on a different computer, but otherwise it sounds good.

i wasn't sure what `--icu` really does. i get that it's some lib for char conversion and it bulks up the bin size quite a bit, but i don't really understand what the real world consequence is. in what way might it affect performance or behavior?

also, do i need `--cmake`? i'm using it headless only. only using the webui and other things that use its api. i saw the hint about qt6. i'm not even using qt, right?

thanks

**Comment by userdocs** [comment]
ICU is the proper, preferred dependency of QT. It's not a qbittorrent dependency. There was a bug before that needed ICU to resolve but I think its's fixed now. If binary size is an issue you can safely not use ICU otherwise just leave this default.

cmake is required to build QT6 and optionally other items. The current qbittorrent project is defaulted to: Qt6 + Libtorrent v1.2

They provide 4 builds in total, similar to what this script is capable to outputting.

As you see when you run the script these are the build defaults

![image](https://github.com/userdocs/qbittorrent-nox-static/assets/16525024/9c8562bb-b3dd-47ce-be53-d3aca4fec454)

So if you don't care about using Qt6 then just use the `-qm` flag and optionally disable ICU and you are good to go.

If you want to use Qt6 it will default to using cmake for building most things.

**Comment by ftc2** [comment]
thanks.

a few more:

1. `-o` (at least in theory) could give performance benefits, right?

2. `-qm -o` yielded a 59MiB bin. did it build with ICU despite not supplying `-i`? i don't care about the bin size, i'm just curious. so i would have to do something like `docker run --env qbt_skip_icu=yes ...` if i wanted to slim it, right?
3.
> The current qbittorrent project is defaulted to: Qt6 + Libtorrent v1.2

```
 ⬤ Default env settings
...
   qbt_libtorrent_tag="v2.0.9"
```

^ is 1.2 really the default?

**Reply by userdocs** [reply]
1: possibly but I'm not sure it makes any perceivable difference to the end user in this usage.

2: Because I turned it on by default but it should probably be off now there is no bug related to not using it. There is no slim build in the context of ICU, there is no specific gain or benefit from the binary being smaller. It's not a performance tweak. It's not better for being smaller, it's not more optimised. It's simply defaulting to a legacy dependency for QT, iconv.

3: It's the default of the qbittorrent project. That default make no difference to this script since it builds and releases both and end users can build what they want.

**Reply by userdocs** [reply]
I also disabled `icu` by default for the script. This returns it to previous behaviour and the `-i` toggle makes more sense now.

**Comment by ftc2** [comment]
unsolicited thoughts on the documentation as a noob...

in [docker build docs](https://github.com/userdocs/qbittorrent-nox-static/blob/master/docs/script-installation.md#docker-via-ssh), if alpine's the preferred platform, why not put it first instead of last? :)

![](https://i.imgur.com/SH9Kjf8.png)

as for the tips just below it, some possible improvements for the clueless (like me):

https://github.com/userdocs/qbittorrent-nox-static/blob/master/docs/script-installation.md?plain=1#L100

> [!NOTE|iconVisibility:hidden|labelVisibility:hidden|style:callout] Please see the [switches and flags summary](/build-help?id=switches-and-flags-summarised) to see what options you can pass and how to use them
You can modify the installation command by editing this part of the docker command.

```bash
bash -s all
```

For example, to use `ICU` using `-i`, `-c` to trigger cmake and in turn activate Qt6 and optimise for the system CPU using `-o`:

```bash
bash -s all -i -c -o
```

- `-o` may give minor performance benefits, but you should only use it if you run qbittorrent on a machine with the same number of cores as the one you build on.
- ICU is the proper, preferred dependency of Qt, but it increases the size of the binary considerably (by ~40MiB). If you don't care about Qt (you are running headless/WebUI-only), you can try `docker run --env qbt_skip_icu=yes ...`
- perhaps also some note about `-i` vs `--env qbt_skip_icu`. i thought that was a little confusing considering it seems to build with ICU regarless of whether or not `-i` is supplied.

**Reply by userdocs** [reply]
The docs are being redone to add info since the v2 of the script. i'll take this into account, thanks.
